service/frontend/task_reachability.go

```diff
@@ -0,0 +1,368 @@
+package frontend
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/api/matchingservice/v1"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+)
+
+// Little helper to concurrently map a function over input and fail fast on error.
+func raceMap[IN any, OUT any](input []IN, mapper func(IN) (OUT, error)) ([]OUT, error) {
```

Comment:
Not sure if this is idiomatic but I found myself repeating this pattern a few times in this file and ended up refactoring to use this little helper. Maybe we have something similar in the codebase or a shared place we can put this.